java-checks/src/main/java/org/sonar/java/locks/LockedVisitor.java

```diff
@@ -0,0 +1,121 @@
+  private static final MethodInvocationMatcher UNLOCK_INVOCATION = MethodInvocationMatcher.create().typeDefinition(TypeCriteria.subtypeOf(JAVA_LOCK)).name("unlock");
+
+  public LockedVisitor(SubscriptionBaseVisitor check) {
+    super(check);
+  }
+
+  private static MethodInvocationMatcherCollection lockMethodInvocationMatcher() {
+    return MethodInvocationMatcherCollection.create(
+      MethodInvocationMatcher.create()
+        .typeDefinition(TypeCriteria.subtypeOf(JAVA_LOCK))
+        .name("lock"),
+      MethodInvocationMatcher.create()
+        .typeDefinition(TypeCriteria.subtypeOf(JAVA_LOCK))
+        .name("lockInterruptibly"),
+      MethodInvocationMatcher.create()
+        .typeDefinition(TypeCriteria.subtypeOf(JAVA_LOCK))
+        .name("tryLock")
+        .withNoParameterConstraint());
+  }
+
+  @Override
+  public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+    ExpressionTree variable = tree.variable();
+    if (variable.is(Tree.Kind.IDENTIFIER, Tree.Kind.MEMBER_SELECT)) {
+
+      IdentifierTree identifier;
+      if (variable.is(Tree.Kind.IDENTIFIER)) {
+        identifier = (IdentifierTree) variable;
+      } else {
+        identifier = ((MemberSelectExpressionTree) variable).identifier();
+      }
+      Symbol symbol = identifier.symbol();
+      if (symbol.type().isSubtypeOf(JAVA_LOCK)) {
+        executionState.newValueForSymbol(symbol, tree, LockState.NULL);
+      }
+    }
+  }
+
+  @Override
+  public void visitMethodInvocation(MethodInvocationTree tree) {
+    if (LOCK_INVOCATIONS.anyMatch(tree)) {
+      Symbol symbol = extractInvokedOnSymbol(tree.methodSelect());
+      if (symbol != null) {
+        executionState.newValueForSymbol(symbol, tree, LockState.LOCKED);
+      }
+    } else if (UNLOCK_INVOCATION.matches(tree)) {
+      ExpressionTree methodSelect = tree.methodSelect();
+      if (methodSelect.is(Tree.Kind.MEMBER_SELECT)) {
+        ExpressionTree expression = ((MemberSelectExpressionTree) methodSelect).expression();
+        if (expression.is(Tree.Kind.IDENTIFIER)) {
+          executionState.markValueAs(((IdentifierTree) expression).symbol(), LockState.UNLOCKED);
+        }
+      }
+    }
+  }
+
+  @CheckForNull
+  private Symbol extractInvokedOnSymbol(ExpressionTree expressionTree) {
+    if (expressionTree.is(Tree.Kind.MEMBER_SELECT)) {
+      return extractSymbol(((MemberSelectExpressionTree) expressionTree).expression());
+    } else if (expressionTree.is(Tree.Kind.IDENTIFIER)) {
+      return ((IdentifierTree) expressionTree).symbol().owner();
+    }
+    return null;
+  }
+
+  @CheckForNull
+  private Symbol extractSymbol(ExpressionTree expressionTree) {
+    if (expressionTree.is(Tree.Kind.MEMBER_SELECT)) {
+      return ((MemberSelectExpressionTree) expressionTree).identifier().symbol();
+    } else if (expressionTree.is(Tree.Kind.IDENTIFIER)) {
+      return ((IdentifierTree) expressionTree).symbol();
+    }
+    return null;
+  }
+
+  public void insertIssues() {
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)Add the "@Override" annotation above this method signature[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1161)